### PR TITLE
add module odgi/layout

### PR DESCRIPTION
**Pull request by subwaystation**
## PR checklist

This PR progresses https://github.com/nf-core/pangenome/issues/70.

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [x] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [x] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

Adding a new module which is able to generate a 2D layout of a pangenome graph with `odgi layout`.

### CODE REVIEW

**Review by nvnieuwk — CHANGES_REQUESTED** [review]
How exactly can you specify the output format?

Can you also add a test that outputs a TSV file?

File: modules/nf-core/odgi/layout/main.nf
Comment (nvnieuwk) [inline modules/nf-core/odgi/layout/main.nf:29]:
```suggestion
        --out ${prefix}.lay \\
        --tsv ${prefix}.tsv \\
        $args
```

File: tests/modules/nf-core/odgi/layout/nextflow.config
Comment (nvnieuwk) [inline tests/modules/nf-core/odgi/layout/nextflow.config:8]:
```suggestion
```

**Review by nvnieuwk — CHANGES_REQUESTED** [review]
The output files should be created in the module itself using the prefix

File: modules/nf-core/odgi/layout/main.nf
Comment (subwaystation) [reply modules/nf-core/odgi/layout/main.nf:29]: But the output is optional. It could the module should only output a TSV and now LAY.
Creating the files directly in the module would mean we always create both files.

File: modules/nf-core/odgi/layout/main.nf
Comment (subwaystation) [reply modules/nf-core/odgi/layout/main.nf:16]: Since either lay, tsv, or both outputs are possible, we have to keep these channels optional.

File: modules/nf-core/odgi/layout/main.nf
Comment (nvnieuwk) [reply modules/nf-core/odgi/layout/main.nf:29]: That's okay, other modules do this too. It's up to the pipeline developer to choose what to save :)

File: modules/nf-core/odgi/layout/main.nf
Comment (nvnieuwk) [reply modules/nf-core/odgi/layout/main.nf:16]: Optional output means that the specified files aren't always part of the process. Since the tool will now create both files every time, it will always be there and isn't optional anymore

File: modules/nf-core/odgi/layout/main.nf
Comment (subwaystation) [reply modules/nf-core/odgi/layout/main.nf:29]: To give you a concrete example:

I have a pangenome graph and want to visualize it in 2D. So I do `odgi layout` with LAY output. Then I use `odgi draw` and have a 2D PNG of the graph.
Or I want to interactively explore the graph with e.g. https://github.com/chfi/gfaestus. Then I would only require the TSV output of `odgi layout`.

It is really rare I want both. We want to save disk space and IO here.

File: modules/nf-core/odgi/layout/main.nf
Comment (subwaystation) [reply modules/nf-core/odgi/layout/main.nf:16]: Yes, but I want these to stay optional :)

File: modules/nf-core/odgi/layout/main.nf
Comment (nvnieuwk) [reply modules/nf-core/odgi/layout/main.nf:29]: Yeah I understand what you mean! Maybe ask a second opinion on this, but I've always done it this way :)

**Comment by subwaystation** [comment]
@nvnieuwk With my explanations, this PR is fine? ;)

**Review by JoseEspinosa — COMMENTED** [review]
If you want to make the output optional, you can take an approach similar to the muscle module, see [here](https://github.com/nf-core/modules/blob/5e34754d42cd2d5d248ca8673c0a53cdf5624905/modules/nf-core/muscle/main.nf#L30-L36). I agree that especially if the files are big enough it is better just to produce one of them provided that this can be controlled by the tool arguments.

**Comment by subwaystation** [comment]
That's a clever way. I will add accordingly.

**Comment by subwaystation** [comment]
But then how to remove the arguments already in `$args`? Or is muscle able to ignore these?

**Comment by subwaystation** [comment]
Assuming I have `--tsv` in args then I would have later in the cli:
```--tsv --tsv PREFIX.tsv
```

How would you solve this?

**Comment by nvnieuwk** [comment]
That's mostly the reason why I wouldn't bother in using `args`. What's the size of the created files usually? Is it a huge problem if they both were created?

**Comment by subwaystation** [comment]
There is no usual file size in pangenomics. For a 92 haplotypes HLA pangenome graph the file sizes are:
```
 - .og: 105MB
 - .lay:  8MB
 - .tsv: 20MB
 ```

However, the ODGI file for a full human pangenome graph from 92 haplotypes of all chromosomes is 120GB. So for this specific data set the TSV could be ~30GB. In the future we want to build graphs from ~300 haplotypes. So yes, the IO will be significant.

I think it is a huge problem that they are both created. And I think a module's functionality should stay as close to the underlying tool as possible.

**Comment by nvnieuwk** [comment]
The prefix should always be specified inside of the config file and should default to `meta.id`

File: modules/nf-core/odgi/layout/meta.yml
Comment (d4straub) [inline modules/nf-core/odgi/layout/meta.yml:27]:
```suggestion
  - prefix:
      type: string
      description: Prefix for output files

```

**Comment by subwaystation** [comment]
> The prefix should always be specified inside of the config file and should default to `meta.id`

@nvnieuwk But that's exactly not possible here. I need the prefix outside of the process. Luckily you said `should` ;)

**Comment by nvnieuwk** [comment]
> @nvnieuwk But that's exactly not possible here. I need the prefix outside of the process. Luckily you said `should` ;)

After looking at this a bit closer, I think you don't even need a prefix. You can access the meta inside of the `args` and specify the prefix inside of the `args`

File: tests/modules/nf-core/odgi/layout/test.yml
Comment (nvnieuwk) [inline tests/modules/nf-core/odgi/layout/test.yml:10]:
```suggestion
```

**Review by nvnieuwk — CHANGES_REQUESTED** [review]
I'll approve after these changes if you agree :)

**Comment by nvnieuwk** [comment]
With these changes you can adjust the prefix inside `args` yourself

**Review by nvnieuwk — APPROVED** [review]

**Comment by nvnieuwk** [comment]
Awesome! Hit it :partying_face:

**Pull request merged**
